# This is the question 3 of Assignment 
#### by Ablokit Joshi (200030)

In [58]:
import cv2 as cv
import numpy as np

In [59]:
img1 = cv.imread('./1.jpg')
img2 = cv.imread('./2.jpg')
img3 = cv.imread('./3.jpg')

Location of Belna and Ambali respectively in three different times given

In [60]:
loc1 = [[240,114],
        [548,181]]
loc2 = [[224, 122],
        [540, 189]]
loc3 = [[217, 134], 
        [529, 197]]


Straight line equation i.e. y = mx+c (where m is slope and c is y-intercept) for Belna and Ambali respectively

In [61]:
slope_B = (loc3[0][1] - loc2[0][1])/(loc3[0][0] - loc2[0][0])
y_inter_B = loc2[0][1] - slope_B*loc2[0][0]

slope_A = (loc3[1][1] - loc2[1][1])/(loc3[1][0] - loc2[1][0])
y_inter_A = loc2[1][1] - slope_A*loc2[1][0]

Intersection of two straight lines

In [62]:
def lines(A,B):
    x = A[1]-B[1]
    y = B[0]-A[0]
    z = A[1]*B[0]-B[1]*A[0]
    return x,y,z

def intersecting_pts(A,B): #solving two equation with two variables using determinant method
    d  = A[0] * B[1] - A[1] * B[0]
    dx = A[2] * B[1] - A[1] * B[2]
    dy = A[0] * B[2] - A[2] * B[0]
    
    if d:
        x = dx / d
        y = dy / d
        return x,y
    else:
        return False

A = lines(loc2[0],loc3[0])
B = lines(loc2[1],loc3[1])
print(intersecting_pts(A,B))

(-76.72368421052632, 637.5263157894736)


Thus they intersect outside the figure

## Pixel Coordinates of given cities and islands
 ### Cities
 ###### Antsiranana = (201, 268)
 ###### Antananarivo = (140, 437)
 ###### Benonoka = (95, 618)
 
 ### Islands
 ###### Mauritius = (416, 476)
 ###### Reunion => (363, 499)

### Let distance in Pixel coordinates be d_p and distance in real coordinates be d_r, then:
    d_r = n*d_p where n is a propotionality constant 

In [63]:
#distance between Antsiranana and  Benonoka in pixel coordinates
dist_c_p = np.sqrt((201-95)**2 + (268-618)**2)

#distance between Mauritius and  Reunion in pixel coordinates
dist_i_p = np.sqrt((416-363)**2 + (476-499)**2)

print("distance between Antsiranana and  Benonoka is: " + str(dist_c_p) + " and distance between Mauritius and  Reunion is: " + str(dist_i_p))
# print(dist_c_p)
# print("\ndistance between Mauritius and  Reunion is: ")
# print(dist_i_p)

distance between Antsiranana and  Benonoka is: 365.6993300513415 and distance between Mauritius and  Reunion is: 57.77542730261716


### Now to find the value of n

In [64]:
n1 = 1500/dist_c_p
n2 = 250/dist_i_p
n = 0.5*(n1+n2) #taking mean of the two
print("Thus the value of n is",n)

Thus the value of n is 4.214414673351662


### Distance of apojuncture from the capital city of Antananarivo

In [65]:
dist_Apo2Anta_pix = np.sqrt((140+76.72368421052632)**2 + (437 - 637.5263157894736)**2)
dist_Apo2Anta_real = n*dist_Apo2Anta_pix
print(dist_Apo2Anta_real)

1244.3587175119246


### Thus the distance between apojuncture and Antananarivo is 1244.36 km

### Distance of apojuncture from last captured cyclones

In [66]:
dist_Apo2Belna = n*(np.sqrt((217+76.72368421052632)**2 + (134 - 637.5263157894736)**2))

dist_Apo2Ambali = n*(np.sqrt((529+76.72368421052632)**2 + (197 - 637.5263157894736)**2))

print("Distance of apojuncture from Belna is " + str(dist_Apo2Belna) + " km and from Ambali is " + str(dist_Apo2Ambali) + " km")

Distance of apojuncture from Belna is 2456.7267057283466 km and from Ambali is 3156.4942288100724 km


### Avg velocity of cyclone Belna

In [67]:
dist_B_loc1to2 = n*(np.sqrt((loc1[0][0] - loc2[0][0])**2 + (loc1[0][1] - loc2[0][1])**2))
time_B_loc1to2 = 41/60 #calculating in hours || 14:03-->14:44
velocity_B_1to2 = dist_B_loc1to2/time_B_loc1to2

dist_B_loc2to3 = n*(np.sqrt((loc2[0][0] - loc3[0][0])**2 + (loc2[0][1] - loc3[0][1])**2))
time_B_loc2to3 = 55/60 #calculating in hours || 14:44-->15:39
velocity_B_2to3 = dist_B_loc2to3/time_B_loc2to3

velocity_B = 0.5*(velocity_B_1to2 + velocity_B_2to3)
print("Avg velocity of cyclone Belna is " + str(velocity_B) + " kmph")

Avg velocity of cyclone Belna is 87.09878178500512 kmph


### Avg velocity of cyclone Ambali

In [68]:
dist_A_loc1to2 = n*(np.sqrt((loc1[1][0] - loc2[1][0])**2 + (loc1[1][1] - loc2[1][1])**2))
time_A_loc1to2 = 41/60 #calculating in hours || 14:03-->14:44
velocity_A_1to2 = dist_A_loc1to2/time_A_loc1to2

dist_A_loc2to3 = n*(np.sqrt((loc2[1][0] - loc3[1][0])**2 + (loc2[1][1] - loc3[1][1])**2))
time_A_loc2to3 = 55/60 #calculating in hours || 14:44-->15:39
velocity_A_2to3 = dist_A_loc2to3/time_A_loc2to3

velocity_A = 0.5*(velocity_A_1to2 + velocity_A_2to3)
print("Avg velocity of cyclone Ambali is " + str(velocity_A) + " kmph")

Avg velocity of cyclone Ambali is 66.15496182021445 kmph


### Time taken for Belna and Ambali to reach apojuncture

In [69]:
time_Belna = float("{:.3f}".format((dist_Apo2Belna/velocity_B)))

time_Ambali = float("{:.3f}".format((dist_Apo2Ambali/velocity_A)))


print("Time taken for Belna to reach apojuncture is " + str(time_Belna) + "hrs and for Ambali is " + str(time_Ambali) + " hrs")

Time taken for Belna to reach apojuncture is 28.206hrs and for Ambali is 47.714 hrs


# SUMMARY

### Distance between apojuncture and Antananarivo is 1244.36 km
### Distance of apojuncture from Belna is 2456.73 km and from Ambali is 3156.49 km
### Avg velocity of cyclone Belna is 87.10 kmph
### Avg velocity of cyclone Ambali is 66.15 kmph
### Time taken for Belna to reach apojuncture is 28.206hrs and for Ambali is 47.714 hrs and thus they will never collide as the time taken is different for both

In [70]:
#End of code